src/native/corehost/corehost.cpp

```diff
@@ -93,6 +100,44 @@ void need_newer_framework_error()
 
 #if defined(CURHOST_EXE)
 
+int try_run_as_arch(const pal::char_t *path, int rc, const int argc, const pal::char_t* argv[])
+{
+    assert((rc == StatusCode::CorehostMustRunAsArchArm64) || (rc == StatusCode::CorehostMustRunAsArchX64));
+#if !defined(FEATURE_APPHOST) && defined(TARGET_OSX)
+
+    trace::verbose(_X("Restarting process in arch %s."), (rc == CorehostMustRunAsArchArm64) ? _X("Arm64") : _X("x64"));
+
+    posix_spawnattr_t attr;
+    int result = 0;
+    size_t size;
+
+    int curproc_arch_affinity = (rc == CorehostMustRunAsArchArm64) ? 0 : 1;
+    size = sizeof(curproc_arch_affinity);
+    sysctlbyname("kern.curproc_arch_affinity", NULL, NULL, &curproc_arch_affinity, size);
+    // kern.curproc_arch_affinity set ignore errors (not available on x64 platform)
+
+    result = posix_spawnattr_init(&attr);
+
+    if (result == 0)
+        result = posix_spawnattr_setflags(&attr, POSIX_SPAWN_SETEXEC);
+
+    if (result == 0)
+    {
+        cpu_type_t cpu_pref = (rc == CorehostMustRunAsArchArm64) ? CPU_TYPE_ARM64 : CPU_TYPE_X86_64;
+        result = posix_spawnattr_setbinpref_np(&attr, 1, &cpu_pref, &size);
+    }
+
+    if (result == 0)
+        result = posix_spawn(NULL, path, NULL, &attr, (pal::char_t * const *)argv, environ);
+
+    trace::error(_X("try_run_as_arch(%s) failed: %s"), path, strerror(errno));
+#else // !defined(FEATURE_APPHOST) && defined(TARGET_OSX)
+    trace::error(_X("try_run_as_arch() unsuppported"), path);
```

Comment:
I think this needs a better error.
If I'm not mistaken this would happen when trying to run x64 6.0 apphost app on Apple Silicon. That's a very possible user scenario and thus should produce somewhat actionable error message. I think we need to ask the PMs what should be the experience in this case (and have a test for it if possible).